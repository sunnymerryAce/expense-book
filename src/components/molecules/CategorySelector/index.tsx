import React, { useState, useEffect } from "react";
import Box from "@material-ui/core/Box";
import CategoryButton, {
  CategoryButtonProps,
} from "@/components/atoms/CategoryButton";
import { createStyles, Theme, makeStyles } from "@material-ui/core/styles";
import { Category, Categories } from "@/state/categories";

interface CategorySelectorProps {
  categories: Categories;
  handleChangeCategory: (...props: any[]) => any;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: "flex",
      justifyContent: "center",
      flexWrap: "wrap",
      "& > *": {
        margin: theme.spacing(0.5),
      },
    },
  })
);

const CategorySelector: React.FC<CategorySelectorProps> = (props) => {
  const classes = useStyles();
  const [selectedCategoryId, setSelectedCategoryId] = useState<string>("");

  useEffect(() => {
    // 初期表示時、先頭のカテゴリを指定
    setSelectedCategoryId(Object.keys(props.categories)[0]);
  }, []);

  const categoryButtonProps = (
    category: Category,
    categoryId: string
  ): CategoryButtonProps => {
    return {
      selected: categoryId === selectedCategoryId,
      label: category.name,
      handleClick: (): void => {
        setSelectedCategoryId(categoryId);
        props.handleChangeCategory(categoryId);
      },
    };
  };

  return (
    <Box className={classes.root}>
      {Object.entries(props.categories).map(([categoryId, category]) => {
        return (
          <CategoryButton
            key={categoryId}
            {...categoryButtonProps(category, categoryId)}
          ></CategoryButton>
        );
      })}
    </Box>
  );
};

export default CategorySelector;
