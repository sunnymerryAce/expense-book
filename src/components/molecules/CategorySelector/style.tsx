import { makeStyles, createStyles, Theme } from "@material-ui/core/styles";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: "flex",
      justifyContent: "center",
      flexWrap: "wrap",
      "& > *": {
        margin: theme.spacing(0.5),
      },
    },
  })
);

export default useStyles;
