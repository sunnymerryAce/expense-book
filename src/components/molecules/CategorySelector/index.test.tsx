import React from "react";
import CategorySelector from "./";
import { shallow, mount } from "enzyme";

describe("<CategorySelector />", () => {
  let container: any = null;
  const categories = {
    aaaaa: {
      defaultAmount: 30000,
      name: "Food",
      sortIndex: 0,
    },
    bbbbb: {
      defaultAmount: 5000,
      name: "Cafe",
      sortIndex: 1,
    },
    ccccc: {
      defaultAmount: 12000,
      name: "雑費",
      sortIndex: 2,
    },
    ddddd: {
      defaultAmount: 45000,
      name: "Drink",
      sortIndex: 3,
    },
    eeeee: {
      defaultAmount: 20000,
      name: "Date",
      sortIndex: 4,
    },
    fffff: {
      defaultAmount: 3000,
      name: "Book",
      sortIndex: 5,
    },
    ggggg: {
      defaultAmount: 12000,
      name: "Gym",
      sortIndex: 6,
    },
    hhhhh: {
      defaultAmount: 33000,
      name: "Fixed",
      sortIndex: 7,
    },
    iiiii: {
      defaultAmount: 30000,
      name: "Sudden",
      sortIndex: 8,
    },
    jjjjj: {
      defaultAmount: 45000,
      name: "Savings",
      sortIndex: 9,
    },
  };

  const props = {
    categories: categories,
    handleChangeCategory: jest.fn(),
    selectedCategoryId: "",
  };

  afterEach(() => {
    container.unmount();
    container = null;
  });

  test("should match the snapshot", () => {
    container = shallow(<CategorySelector {...props} />);
    expect(container.html()).toMatchSnapshot();
  });

  test("should have proper length of CategoryButton", () => {
    container = shallow(<CategorySelector {...props} />);
    expect(container.find("CategoryButton").getElements().length).toBe(
      Object.keys(props.categories).length
    );
  });

  test("should have selected first category", () => {
    container = mount(<CategorySelector {...props} />);
    expect(container.find(".CategoryButton").at(0).props().color).toBe(
      "secondary"
    );
  });

  test("should change selected category", () => {
    container = mount(<CategorySelector {...props} />);
    container.find(".CategoryButton").at(1).simulate("click");
    expect(container.find(".CategoryButton").at(1).props().color).toBe(
      "secondary"
    );
  });

  test("event handler should be triggered", () => {
    container = mount(<CategorySelector {...props} />);
    container.find(".CategoryButton").at(1).simulate("click");
    expect(props.handleChangeCategory).toBeCalled();
  });
});
