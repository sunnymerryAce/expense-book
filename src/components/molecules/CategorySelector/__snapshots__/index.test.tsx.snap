// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategorySelector /> should match the snapshot 1`] = `"<div class=\\"MuiBox-root MuiBox-root-2 makeStyles-root-1\\"><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorSecondary MuiChip-clickableColorSecondary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Food</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Cafe</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">雑費</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Dating</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Free</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Book</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Gym</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Fixed</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Sudden</span></div><div class=\\"MuiButtonBase-root MuiChip-root CategoryButton MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable\\" tabindex=\\"0\\" role=\\"button\\"><span class=\\"MuiChip-label\\">Saving</span></div></div>"`;
